tests/functions/test_expand_grid.py

```diff
@@ -0,0 +1,635 @@
+def compute_two_dfs(df1, df2):
+    #get lengths of dataframes(number of rows) and swap
+    # essentially we'll pair one dataframe with the other's length:
+    lengths = reversed([ent.index.size for ent in (df1, df2)])
+    #grab the maximum string length
+    string_cols = [
+        frame.select_dtypes(include="object").columns for frame in (df1, df2)
+    ]
+
+    #pair max string length with col
+    #will be passed into frame.to_records, to get dtype in numpy recarray
+    string_cols = [{col: f"<U{frame[col].str.len().max()}"
+                    for col in ent}
+                   for ent, frame in zip(string_cols, (df1, df2))]
+
+    (len_first, col_dtypes,
+     first), (len_last, col_dtypes,
+              last) = list(zip(lengths, string_cols, (df1, df2)))
+
+    #export to numpy as recarray
+    first = first.to_records(column_dtypes=col_dtypes, index=False)
+    #tile first with len_first
+    #remember, len_first is the length of the other dataframe
+    first = np.tile(first, (len_first, 1))
+    #get a 1d array
+    first = np.concatenate(first)
+    #sorting here ensures we get each row of the first
+    #with the entire rows of the other dataframe
+    np.recarray.sort(first, order=first.dtype.names[0])
+
+    #same process as first, except there'll be no sorting
+    last = last.to_records(column_dtypes=col_dtypes, index=False)
+    last = np.tile(last, (len_last, 1))
+    last = np.concatenate(last)
+    result = rfn.merge_arrays((first, last), flatten=True, asrecarray=True)
+    return pd.DataFrame.from_records(result)
+
+
+def grid_computation_list(dfs):
+    return reduce(compute_two_dfs, dfs)
+
+
+def grid_computation(dfs, dicts):
+    if not dicts:
+        result = grid_computation_list(dfs)
+    elif not dfs:
+        result = grid_computation_dict(dicts)
+    else:
+        dfs.append(grid_computation_dict(dicts))
+        result = grid_computation_list(dfs)
+    return result
+
+
+########################### Tests
+
+
+def test_not_a_dict():
+    """Test that entry(list) is not a dictionary"""
+    data = [60, 70]
+    with pytest.raises(ValueError):
+        assert expand_grid(data)
+
+
+def test_not_a_dict_1():
+    """Test that entry (dataframe) is not a dictionary"""
+    data = pd.DataFrame([60, 70])
+    with pytest.raises(ValueError):
+        assert expand_grid(data)
+
+
+def test_empty_dict():
+    """Test that entry should not be empty"""
+    data = {}
+    with pytest.raises(ValueError):
+        assert expand_grid(data)
+
+
+def test_scalar_to_list():
+    """Test that 
+    scalars(null,boolean,int,float), strings, sets, and tuples 
+    are converted to lists in the dictionary values
+     """
+    data = {
+        "x": 1,
+        "y": "string",
+        "z": set((2, 3, 4)),
+        "a": tuple((26, 50)),
+        "b": None,
+        "c": 1.2,
+        "d": True,
+        "e": False
+    }
+    expected = ([], {
+        'x': [1],
+        'y': ['s', 't', 'r', 'i', 'n', 'g'],
```

Comment:
Can I ask, is this the only intended behaviour? Though it is idiomatic from a Python programming perspective, I wonder if an end-user might be surprised, having expected `["string"]` as the output instead? What are your thoughts on this? 

We can keep things as is, though because this might be a user-facing behaviour issue, our final decision should go into the docstring that explains the behaviour.

Reply:
that's my assumption, that if a string is passed in, then it will be chunked into individual alphabets. I'll probably run a test in R and see how strings are executed in expand_grid. You are more experienced with R than I am though, so I'll defer to you.